mmcls/models/backbones/repvgg.py

```diff
@@ -441,7 +490,7 @@ def __init__(self,
         self.with_cp = with_cp
         self.norm_eval = norm_eval
 
-        channels = min(64, int(base_channels * self.arch['width_factor'][0]))
+        channels = min(32, int(base_channels * self.arch['width_factor'][0]))
```

Comment:
Ok, I will implement arch='small' for YOLOX-PAI-small.

But for YOLOX-PAI-small, stem-channels should be set to 32. How about this?

```
min_stem_channels = 64
...
channels = min(min_stem_channels, int(base_channels * self.arch['width_factor'][0]))
```

Reply:
Ok, I will implement arch='small' for YOLOX-PAI-small.

But for YOLOX-PAI-small, stem-channels should be set to 32. How about this?

```
min_stem_channels = 64
...
channels = min(min_stem_channels, int(base_channels * self.arch['width_factor'][0]))
```

Reply:
I check the ordinary [imp](https://github.com/alibaba/EasyCV/blob/9f01a37ad4df57b30430c41df08459025174e8fd/easycv/models/backbones/repvgg_yolox_backbone.py#L310) and that in [mmdet PR](https://github.com/RistoranteRist/mmdetection/blob/55d1dfefc539a08ce74b7756b05af178ab53c161/configs/yoloxpai/yolox_pai_s_8x8_300e_coco.py#L11) of YOLOX-PAI, the `channels` is 
```
[64, 128, 256, 512, 1024]
```

**NOTE：  I'm not sure if there is an mistake here.**
#### the backbone in mmdet
```
 backbone=dict(
        _delete_=True,
        type='mmcls.RepVGG',
        arch=dict(
            num_blocks=[3, 5, 7, 3],
            base_channels=32,  # no use, the `base_channels` in backbone defaults to be 64.
            width_factor=[1, 1, 1, 1], # 64, 128, 256, 512
            group_layer_map=None,
            se_cfg=None),
        add_ppf=True,
        norm_cfg=dict(type='BN', eps=0.001, momentum=0.03),
        out_indices=(1, 2, 3),
    ),
```
in-channels and out-channels (stem, stage1, stage2, stage3, stahge4) should be [(3, 64), (64, 64), (64, 128), (128, 256), (256, 512)].

Is what your want the following?
in-channels and out-channels (stem, stage1, stage2, stage3, stahge4) should be [(3, 32), (32, 64), (64, 128), (128, 256), (256, 512)].

Reply:
Actually I used custom RepVGG for YOLOX-PAI and I set `base_channels=32`.

```
import warnings

import torch
from mmcls.models.backbones.repvgg import RepVGG as Base
from mmcls.models.backbones.repvgg import RepVGGBlock
from mmcv.cnn import ConvModule
from mmdet.models.builder import BACKBONES
from torch import nn


class MTSPPF(nn.Module):
    def __init__(self,
                 in_channels,
                 out_channels,
                 norm_cfg=dict(type='BN'),
                 act_cfg=dict(type='ReLU'),
                 kernel_size=5):
        super().__init__()
        c_ = in_channels // 2  # hidden channels
        self.cv1 = ConvModule(in_channels,
                              c_,
                              1,
                              stride=1,
                              norm_cfg=norm_cfg,
                              act_cfg=act_cfg)
        self.cv2 = ConvModule(c_ * 4,
                              out_channels,
                              1,
                              stride=1,
                              norm_cfg=norm_cfg,
                              act_cfg=act_cfg)
        self.maxpool = nn.MaxPool2d(kernel_size=kernel_size,
                                    stride=1,
                                    padding=kernel_size // 2)

    def forward(self, x):
        x = self.cv1(x)
        with warnings.catch_warnings():
            warnings.simplefilter('ignore')
            y1 = self.maxpool(x)
            y2 = self.maxpool(y1)
            return self.cv2(torch.cat([x, y1, y2, self.maxpool(y2)], 1))


@BACKBONES.register_module(force=True)
class RepVGG(Base):
    """RepVGG backbone.
    A PyTorch impl of : `RepVGG: Making VGG-style ConvNets Great Again
    <https://arxiv.org/abs/2101.03697>`_
    Args:
        arch (str | dict): The parameter of RepVGG.
            If it's a dict, it should contain the following keys:
            - num_blocks (Sequence[int]): Number of blocks in each stage.
            - width_factor (Sequence[float]): Width deflator in each stage.
            - group_layer_map (dict | None): RepVGG Block that declares
              the need to apply group convolution.
            - se_cfg (dict | None): Se Layer config
        in_channels (int): Number of input image channels. Default: 3.
        base_channels (int): Base channels of RepVGG backbone, work
            with width_factor together. Default: 64.
        out_indices (Sequence[int]): Output from which stages. Default: (3, ).
        strides (Sequence[int]): Strides of the first block of each stage.
            Default: (2, 2, 2, 2).
        dilations (Sequence[int]): Dilation of each stage.
            Default: (1, 1, 1, 1).
        frozen_stages (int): Stages to be frozen (all param fixed). -1 means
            not freezing any parameters. Default: -1.
        conv_cfg (dict | None): The config dict for conv layers. Default: None.
        norm_cfg (dict): The config dict for norm layers.
            Default: dict(type='BN').
        act_cfg (dict): Config dict for activation layer.
            Default: dict(type='ReLU').
        with_cp (bool): Use checkpoint or not. Using checkpoint will save some
            memory while slowing down the training speed. Default: False.
        deploy (bool): Whether to switch the model structure to deployment
            mode. Default: False.
        norm_eval (bool): Whether to set norm layers to eval mode, namely,
            freeze running stats (mean and var). Note: Effect on Batch Norm
            and its variants only. Default: False.
        init_cfg (dict or list[dict], optional): Initialization config dict.
    """

    groupwise_layers = [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26]
    g2_layer_map = {layer: 2 for layer in groupwise_layers}
    g4_layer_map = {layer: 4 for layer in groupwise_layers}

    arch_settings = {
        'A0':
        dict(num_blocks=[2, 4, 14, 1],
             width_factor=[0.75, 0.75, 0.75, 2.5],
             group_layer_map=None,
             se_cfg=None),
        'A1':
        dict(num_blocks=[2, 4, 14, 1],
             width_factor=[1, 1, 1, 2.5],
             group_layer_map=None,
             se_cfg=None),
        'A2':
        dict(num_blocks=[2, 4, 14, 1],
             width_factor=[1.5, 1.5, 1.5, 2.75],
             group_layer_map=None,
             se_cfg=None),
        'B0':
        dict(num_blocks=[4, 6, 16, 1],
             width_factor=[1, 1, 1, 2.5],
             group_layer_map=None,
             se_cfg=None),
        'B1':
        dict(num_blocks=[4, 6, 16, 1],
             width_factor=[2, 2, 2, 4],
             group_layer_map=None,
             se_cfg=None),
        'B1g2':
        dict(num_blocks=[4, 6, 16, 1],
             width_factor=[2, 2, 2, 4],
             group_layer_map=g2_layer_map,
             se_cfg=None),
        'B1g4':
        dict(num_blocks=[4, 6, 16, 1],
             width_factor=[2, 2, 2, 4],
             group_layer_map=g4_layer_map,
             se_cfg=None),
        'B2':
        dict(num_blocks=[4, 6, 16, 1],
             width_factor=[2.5, 2.5, 2.5, 5],
             group_layer_map=None,
             se_cfg=None),
        'B2g2':
        dict(num_blocks=[4, 6, 16, 1],
             width_factor=[2.5, 2.5, 2.5, 5],
             group_layer_map=g2_layer_map,
             se_cfg=None),
        'B2g4':
        dict(num_blocks=[4, 6, 16, 1],
             width_factor=[2.5, 2.5, 2.5, 5],
             group_layer_map=g4_layer_map,
             se_cfg=None),
        'B3':
        dict(num_blocks=[4, 6, 16, 1],
             width_factor=[3, 3, 3, 5],
             group_layer_map=None,
             se_cfg=None),
        'B3g2':
        dict(num_blocks=[4, 6, 16, 1],
             width_factor=[3, 3, 3, 5],
             group_layer_map=g2_layer_map,
             se_cfg=None),
        'B3g4':
        dict(num_blocks=[4, 6, 16, 1],
             width_factor=[3, 3, 3, 5],
             group_layer_map=g4_layer_map,
             se_cfg=None),
        'D2se':
        dict(num_blocks=[8, 14, 24, 1],
             width_factor=[2.5, 2.5, 2.5, 5],
             group_layer_map=None,
             se_cfg=dict(ratio=16, divisor=1))
    }

    def __init__(self,
                 arch,
                 in_channels=3,
                 base_channels=64,
                 out_indices=(3, ),
                 strides=(2, 2, 2, 2),
                 dilations=(1, 1, 1, 1),
                 frozen_stages=-1,
                 conv_cfg=None,
                 norm_cfg=dict(type='BN'),
                 act_cfg=dict(type='ReLU'),
                 with_cp=False,
                 deploy=False,
                 norm_eval=False,
                 add_ppf=False,
                 init_cfg=[
                     dict(type='Kaiming', layer=['Conv2d']),
                     dict(type='Constant',
                          val=1,
                          layer=['_BatchNorm', 'GroupNorm'])
                 ]):
        super(Base, self).__init__(init_cfg)

        if isinstance(arch, str):
            assert arch in self.arch_settings, \
                f'"arch": "{arch}" is not one of the arch_settings'
            arch = self.arch_settings[arch]
        elif not isinstance(arch, dict):
            raise TypeError('Expect "arch" to be either a string '
                            f'or a dict, got {type(arch)}')

        assert len(arch['num_blocks']) == len(
            arch['width_factor']) == len(strides) == len(dilations)
        assert max(out_indices) < len(arch['num_blocks'])
        if arch['group_layer_map'] is not None:
            assert max(arch['group_layer_map'].keys()) <= sum(
                arch['num_blocks'])

        if arch['se_cfg'] is not None:
            assert isinstance(arch['se_cfg'], dict)

        self.arch = arch
        self.in_channels = in_channels
        self.base_channels = base_channels
        self.out_indices = out_indices
        self.strides = strides
        self.dilations = dilations
        self.deploy = deploy
        self.frozen_stages = frozen_stages
        self.conv_cfg = conv_cfg
        self.norm_cfg = norm_cfg
        self.act_cfg = act_cfg
        self.with_cp = with_cp
        self.norm_eval = norm_eval

        channels = min(32, int(base_channels * self.arch['width_factor'][0]))
        self.stem = RepVGGBlock(self.in_channels,
                                channels,
                                stride=2,
                                se_cfg=arch['se_cfg'],
                                with_cp=with_cp,
                                conv_cfg=conv_cfg,
                                norm_cfg=norm_cfg,
                                act_cfg=act_cfg,
                                deploy=deploy)

        next_create_block_idx = 1
        self.stages = []
        for i in range(len(arch['num_blocks'])):
            num_blocks = self.arch['num_blocks'][i]
            stride = self.strides[i]
            dilation = self.dilations[i]
            out_channels = int(base_channels * 2**i *
                               self.arch['width_factor'][i])

            stage, next_create_block_idx = self._make_stage(
                channels, out_channels, num_blocks, stride, dilation,
                next_create_block_idx, init_cfg)
            stage_name = f'stage_{i + 1}'
            self.add_module(stage_name, stage)
            self.stages.append(stage_name)

            channels = out_channels

        if add_ppf:
            self.ppf = MTSPPF(out_channels,
                              out_channels,
                              norm_cfg=norm_cfg,
                              act_cfg=act_cfg,
                              kernel_size=5)
        else:
            self.ppf = None

    def forward(self, x):
        x = self.stem(x)
        outs = []
        for i, stage_name in enumerate(self.stages):
            stage = getattr(self, stage_name)
            x = stage(x)
            if i + 1 == len(self.stages) and self.ppf is not None:
                x = self.ppf(x)
            if i in self.out_indices:
                outs.append(x)

        return tuple(outs)
```

Reply:
Official repo set [width to 0.5](https://github.com/alibaba/EasyCV/blob/master/easycv/models/detection/detectors/yolox/yolox.py#L33).
And [channels_list_backbone was multiplied by width](https://github.com/alibaba/EasyCV/blob/9f01a37ad4df57b30430c41df08459025174e8fd/easycv/models/backbones/repvgg_yolox_backbone.py#L317).
So [(3, 32), (32, 64), (64, 128), (128, 256), (256, 512)] is correct.

Reply:
> min_stem_channels = 64
 ...
 channels = min(min_stem_channels, int(base_channels * self.arch['width_factor'][0]))

Yes, It is OK! 


> arch=dict(
            num_blocks=[3, 5, 7, 3],
            base_channels=32,  
            width_factor=[1, 1, 1, 1], # 64, 128, 256, 512
           group_layer_map=None,
           se_cfg=None),


I also think **`base_channels` is more reasonable to set in `arch`**.

To prevent BC breaking, I think it is possible to add a piece of logic：
1. modify `base_channels=64` to `base_channels=None` in [this line](https://github.com/open-mmlab/mmclassification/blob/8d1bc557ab25ad75c17091711629d684cbdba5b8/mmcls/models/backbones/repvgg.py#L392), it has the highest priority, and defaults to None;
2. try to get `base_channels` from the `arch`, the second highest priority;
3. If 1 and 2 all not set `base_channels`, set `base_channels = 64`， the lowest priority, set when `base_channels=None` and there is no keyword of `base_channels` in an arch.